File: LeetCode Tutorials/Java/Sorts/RecursiveSorts/MergeSort/MergeSort.java
package MergeSort;
import java.util.Arrays;

public class MergeSort {

    // Public wrapper for ascending merge sort
    public static void mergeSortAsc(int[] arr) {
        mergeSortAsc(arr, 0, arr.length - 1);
    }
    
    // Public wrapper for descending merge sort
    public static void mergeSortDesc(int[] arr) {
        mergeSortDesc(arr, 0, arr.length - 1);
    }
    
    // Recursive merge sort method for ascending order (internal use)
    private static void mergeSortAsc(int[] arr, int left, int right) {
        if (left >= right) return;
        int mid = left + (right - left) / 2;
        mergeSortAsc(arr, left, mid);
        mergeSortAsc(arr, mid + 1, right);
        mergeAsc(arr, left, mid, right);
    }

    // Recursive merge sort method for descending order (internal use)
    private static void mergeSortDesc(int[] arr, int left, int right) {
        if (left >= right) return;
        int mid = left + (right - left) / 2;
        mergeSortDesc(arr, left, mid);
        mergeSortDesc(arr, mid + 1, right);
        mergeDesc(arr, left, mid, right);
    }
    
    // Merge two sorted subarrays into ascending order
    private static void mergeAsc(int[] arr, int left, int mid, int right) {
        int[] leftArray = Arrays.copyOfRange(arr, left, mid + 1);
        int[] rightArray = Arrays.copyOfRange(arr, mid + 1, right + 1);
        int i = 0, j = 0, k = left;
        
        while (i < leftArray.length && j < rightArray.length) {
            if (leftArray[i] <= rightArray[j]) {
                arr[k++] = leftArray[i++];
            } else {
                arr[k++] = rightArray[j++];
            }
        }
        
        while (i < leftArray.length) {
            arr[k++] = leftArray[i++];
        }
        
        while (j < rightArray.length) {
            arr[k++] = rightArray[j++];
        }
    }
    
    // Merge two sorted subarrays into descending order
    private static void mergeDesc(int[] arr, int left, int mid, int right) {
        int[] leftArray = Arrays.copyOfRange(arr, left, mid + 1);
        int[] rightArray = Arrays.copyOfRange(arr, mid + 1, right + 1);
        int i = 0, j = 0, k = left;
        
        while (i < leftArray.length && j < rightArray.length) {
            if (leftArray[i] >= rightArray[j]) {
                arr[k++] = leftArray[i++];
            } else {
                arr[k++] = rightArray[j++];
            }
        }
        
        while (i < leftArray.length) {
            arr[k++] = leftArray[i++];
        }
        
        while (j < rightArray.length) {
            arr[k++] = rightArray[j++];
        }
    }
    
    public static void main(String[] args) {
        int[] originalArray = {9, 2, 3, 1, 2, 2, 5, 1, 31, 2};
        System.out.println("Original array: " + Arrays.toString(originalArray));
        
        // Create a copy for ascending sort
        int[] ascArray = Arrays.copyOf(originalArray, originalArray.length);
        mergeSortAsc(ascArray);
        System.out.println("Ascending sort: " + Arrays.toString(ascArray));
        
        // Create a copy for descending sort
        int[] descArray = Arrays.copyOf(originalArray, originalArray.length);
        mergeSortDesc(descArray);
        System.out.println("Descending sort: " + Arrays.toString(descArray));
    }
}

File: LeetCode Tutorials/JavaScript/Sorts/RecursiveSorts/MergeSort.js
function mergeAsc(arr1, arr2) {
  let combined = [];
  let i = 0;
  let j = 0;

  while (i < arr1.length && j < arr2.length) {
    if (arr1[i] < arr2[j]) {
      // for descending order use arr1[i] > arr2[j]
      combined.push(arr1[i]);
      i++;
    } else {
      combined.push(arr2[j]);
      j++;
    }
  }
  while (i < arr1.length) {
    combined.push(arr1[i]);
    i++;
  }
  while (j < arr2.length) {
    combined.push(arr2[j]);
    j++;
  }

  return combined;
}

function mergeDesc(arr1, arr2) {
  let combined = [];
  let i = 0;
  let j = 0;

  while (i < arr1.length && j < arr2.length) {
    if (arr1[i] > arr2[j]) {
      combined.push(arr1[i]);
      i++;
    } else {
      combined.push(arr2[j]);
      j++;
    }
  }
  while (i < arr1.length) {
    combined.push(arr1[i]);
    i++;
  }
  while (j < arr2.length) {
    combined.push(arr2[j]);
    j++;
  }

  return combined;
}

function mergeSortAsc(array) {
  if (array.length === 1) return array;

  let midIndex = Math.floor(array.length / 2);
  let left = mergeSortAsc(array.slice(0, midIndex));
  let right = mergeSortAsc(array.slice(midIndex));

  return mergeAsc(left, right);
}

function mergeSortDesc(array) {
  if (array.length === 1) return array;

  let midIndex = Math.floor(array.length / 2);
  let left = mergeSortDesc(array.slice(0, midIndex));
  let right = mergeSortDesc(array.slice(midIndex));

  return mergeDesc(left, right);
}
originalArray = [4, 2, 3, 1, -1];

console.log(mergeSortAsc(originalArray));

console.log(mergeSortDesc(originalArray));
